Modules/zlibmodule.c

```diff
@@ -10,10 +10,12 @@
 #include "zlib.h"
 
 
-#define ENTER_ZLIB(obj) \
-    Py_BEGIN_ALLOW_THREADS; \
-    PyThread_acquire_lock((obj)->lock, 1); \
-    Py_END_ALLOW_THREADS;
+#define ENTER_ZLIB(obj) do {                      \
+    if (!PyThread_acquire_lock((obj)->lock, 0)) { \
+        Py_BEGIN_ALLOW_THREADS                    \
+        PyThread_acquire_lock((obj)->lock, 1);    \
+        Py_END_ALLOW_THREADS                      \
+    } } while (0)
```

Comment:
You didn't make this change in the backports. Why not?

Reply:
This line above the change was added in 3.10.
`zlibstate *state = PyType_GetModuleState(cls);`

So I created PRs separately (3.10 and 3.9-) for code conflict.

The change to `ENTER_ZLIB(obj)` is not a bug fix, maybe only apply it to the master branch?
The code comes frome bz2/lzma module without any changes:
https://github.com/python/cpython/blob/v3.10.0a7/Modules/_bz2module.c#L22-L28
https://github.com/python/cpython/blob/v3.10.0a7/Modules/_lzmamodule.c#L18-L24

(`zlib` module was added in 1997, `bz2` module was added in 2002, `lzma` module was added in 2011.)

Reply:
OK, makes sense.